mlflow/langchain/runnables.py

```diff
@@ -0,0 +1,229 @@
+            if type(steps) == list:
+                steps.append(runnable)
+            elif type(steps) == dict:
+                steps[step] = runnable
+
+    if model_type == RunnableSequence.__name__:
+        return runnable_sequence_from_steps(steps)
+    if model_type == RunnableParallel.__name__:
+        return RunnableParallel(steps)
+
+
+def runnable_sequence_from_steps(steps):
+    """
+    Construct a RunnableSequence from steps.
+
+    :param steps: List of steps to construct the RunnableSequence from.
+    """
+    from langchain.schema.runnable import RunnableSequence
+
+    if len(steps) < 2:
+        raise ValueError(f"RunnableSequence must have at least 2 steps, got {len(steps)}.")
+
+    if len(steps) == 2:
+        return RunnableSequence(first=steps[0], last=steps[1])
+    return RunnableSequence(first=steps[0], middle=steps[1:-1], last=steps[-1])
+
+
+def _save_runnable_with_steps(steps, file_path: Union[Path, str], loader_fn=None, persist_dir=None):
+    """
+    Save the model.
+
+    :steps: steps of the runnable.
+    :param file_path: Path to file to save the model to.
+    """
+    # Convert file to Path object.
+    save_path = Path(file_path) if isinstance(file_path, str) else file_path
+    save_path.mkdir(parents=True, exist_ok=True)
+
+    if isinstance(steps, list):
+        generator = enumerate(steps)
+    elif isinstance(steps, dict):
+        generator = steps.items()
+    unsaved_runnables = {}
+    steps_conf = {}
+    for key, runnable in generator:
+        step = str(key)
+        steps_conf[step] = {}
+        if isinstance(runnable, lc_runnables_types()):
+            steps_conf[step][_MODEL_TYPE_KEY] = runnable.__class__.__name__
+            steps_conf[step].update(
+                _save_runnables(runnable, save_path, step, loader_fn, persist_dir)
+            )
+        elif isinstance(runnable, base_lc_types()):
+            save_runnable_path = f"{save_path!s}/{step}"
+            mkdir(save_runnable_path)
+            lc_model = _validate_and_wrap_lc_model(runnable, loader_fn)
+            steps_conf[step][_MODEL_TYPE_KEY] = lc_model.__class__.__name__
+            steps_conf[step].update(
+                _save_base_lcs(lc_model, save_runnable_path, loader_fn, persist_dir)
+            )
+        else:
+            steps_conf[step][_MODEL_TYPE_KEY] = runnable.__class__.__name__
+            # Save some simple runnables that langchain natively supports.
+            if hasattr(runnable, "save"):
+                runnable.save(f"{save_path!s}/{step}.yaml")
+            # TODO: check if `dict` is enough to load it back
+            elif hasattr(runnable, "dict"):
+                runnable_dict = runnable.dict()
+                with open(f"{save_path!s}/{step}.yaml", "w") as f:
+                    yaml.dump(runnable_dict, f, default_flow_style=False)
+            else:
+                unsaved_runnables[step] = f"{runnable!r}"
```

Comment:
Can we throw immediately if we find an unserializable runnable?

Reply:
i guess the benefit of not doing that is that the user can see all the types of unsupported runnables

Reply:
makes sense